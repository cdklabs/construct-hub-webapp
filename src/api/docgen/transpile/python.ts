import { toSnakeCase } from "codemaker";
import * as reflect from "jsii-reflect";
import { Markdown } from "../render/markdown";
import * as transpile from "./transpile";

/**
 * A python transpiler.
 */
export class PythonTranspile extends transpile.TranspileBase {
  constructor(private readonly ts: reflect.TypeSystem) {
    super("python");
  }

  public readme(readme: string): string {
    return readme;
  }

  public unionOfTypes(
    refs: transpile.TranspiledTypeReference[]
  ): transpile.TranspiledTypeReference {
    const typing = this.typing("Union");
    return {
      raw: `${typing}[${refs.map((r) => r.raw).join(", ")}]`,
      markdown: `${Markdown.bold(typing)}[${refs
        .map((r) => r.markdown)
        .join(", ")}]`,
    };
  }

  public listOfType(
    ref: transpile.TranspiledTypeReference
  ): transpile.TranspiledTypeReference {
    const typing = this.typing("List");
    return {
      raw: `${typing}[${ref.raw}]`,
      markdown: `${Markdown.bold(typing)}[${ref.markdown}]`,
    };
  }

  public mapOfType(
    ref: transpile.TranspiledTypeReference
  ): transpile.TranspiledTypeReference {
    const typing = this.typing("Mapping");
    return {
      raw: `${typing}[#${ref.raw}]`,
      markdown: `${Markdown.bold(this.typing("Mapping"))}[${ref.markdown}]`,
    };
  }

  public any(): transpile.TranspiledTypeReference {
    const typing = this.typing("Any");
    return { raw: typing, markdown: Markdown.pre(typing) };
  }

  public boolean(): transpile.TranspiledTypeReference {
    const b = "bool";
    return { raw: b, markdown: Markdown.pre(b) };
  }

  public str(): transpile.TranspiledTypeReference {
    const s = "str";
    return { raw: s, markdown: Markdown.pre(s) };
  }

  public number(): transpile.TranspiledTypeReference {
    const types = ["int", "float"];
    const typing = this.typing("Union");
    return {
      raw: `${typing}[${types.join(", ")}]`,
      markdown: `${Markdown.bold(this.typing("Union"))}[${types
        .map((t) => Markdown.pre(t))
        .join(", ")}]`,
    };
  }

  public date(): transpile.TranspiledTypeReference {
    const d = "datetime.datetime";
    return { raw: d, markdown: Markdown.pre(d) };
  }

  public enum(enu: reflect.EnumType): transpile.TranspiledEnum {
    return {
      fqn: this.type(enu).fqn,
      name: enu.name,
    };
  }

  public enumMember(em: reflect.EnumMember): transpile.TranspiledEnumMember {
    return {
      fqn: `${this.enum(em.enumType).fqn}.${em.name}`,
      name: em.name,
    };
  }
  public json(): transpile.TranspiledTypeReference {
    const a = "any";
    return { raw: a, markdown: Markdown.pre(a) };
  }

  public property(property: reflect.Property): transpile.TranspiledProperty {
    return {
      name: property.const ? property.name : toSnakeCase(property.name),
      parentType: this.type(property.parentType),
      typeReference: this.typeReference(property.type),
      optional: property.optional,
    };
  }

  public class(klass: reflect.ClassType): transpile.TranspiledClass {
    return {
      name: klass.name,
      type: this.type(klass),
    };
  }

  public parameter(
    parameter: reflect.Parameter
  ): transpile.TranspiledParameter {
    return {
      name: toSnakeCase(parameter.name),
      parentType: this.type(parameter.parentType),
      typeReference: this.typeReference(parameter.type),
      optional: parameter.optional,
    };
  }

  public struct(struct: reflect.InterfaceType): transpile.TranspiledStruct {
    const type = this.type(struct);
    const inputs = struct.allProperties.map((p) =>
      formatInput(this.property(p))
    );
    return {
      type: type,
      name: struct.name,
      import: formatImport(type),
      initialization: formatInvocation(type, inputs),
    };
  }

  public callable(callable: reflect.Callable): transpile.TranspiledCallable {
    const type = this.type(callable.parentType);

    const parameters = new Array<reflect.Parameter>();

    for (const p of callable.parameters.sort(this.optionalityCompare)) {
      if (!this.isStruct(p)) {
        parameters.push(p);
      } else {
        // struct parameters are expanded to the individual struct properties
        const struct = p.parentType.system.findInterface(p.type.fqn!);
        for (const property of struct.allProperties) {
          const parameter = propertyToParameter(callable, property);
          parameters.push(parameter);
        }
      }
    }

    const name = toSnakeCase(callable.name);
    const inputs = parameters.map((p) => formatInput(this.parameter(p)));

    return {
      name,
      parentType: type,
      import: formatImport(type),
      parameters,
      signature: formatSignature(name, inputs),
      invocation: formatInvocation(
        type,
        inputs,
        callable.kind === reflect.MemberKind.Initializer ? undefined : name
      ),
    };
  }

  public interface(
    iface: reflect.InterfaceType
  ): transpile.TranspiledInterface {
    return {
      name: iface.name,
      type: this.type(iface),
    };
  }

  public type(type: reflect.Type): transpile.TranspiledType {
    const module = this.moduleFqn(type.fqn);

    const fqn = [module];

    if (type.namespace) {
      fqn.push(type.namespace);
    }

    fqn.push(type.name);

    return { fqn: fqn.join("."), moduleFqn: module };
  }

  private moduleFqn(fqn: string): string {
    const type = this.ts.findFqn(fqn);
    const submodule = this.findSubmodule(type);

    const targets = submodule ? submodule.targets : type.assembly.targets;

    if (!targets) {
      throw new Error(`Unable to find 'targets' for fqn ${type.fqn}`);
    }

    if (!targets.python) {
      throw new Error(`Python is not a supported target for fqn ${type.fqn}`);
    }

    return targets.python.module;
  }

  private optionalityCompare(
    p1: reflect.Parameter,
    p2: reflect.Parameter
  ): number {
    if (!p1.optional && p1.optional) {
      return -1;
    }
    if (!p2.optional && p1.optional) {
      return 1;
    }
    return 0;
  }

  private isStruct(p: reflect.Parameter): boolean {
    return p.type.fqn ? p.system.findFqn(p.type.fqn).isDataType() : false;
  }

  private findSubmodule(type: reflect.Type): reflect.Submodule | undefined {
    if (!type.namespace) {
      return undefined;
    }

    // if the type is in a submodule, the submodule name is the first
    // part of the namespace. we construct the full submodule fqn and seach for it.
    const submoduleFqn = `${type.assembly.name}.${
      type.namespace.split(".")[0]
    }`;
    const submodules = type.assembly.submodules.filter(
      (s) => s.fqn === submoduleFqn
    );

    if (submodules.length > 1) {
      // can never happen, but the array data structure forces this handling.
      throw new Error(`Found multiple submodulues with fqn ${submoduleFqn}`);
    }

    if (submodules.length === 0) {
      return undefined;
    }

    // type is inside this submodule.
    return submodules[0];
  }

  private typing(type: "List" | "Mapping" | "Any" | "Union"): string {
    return `typing.${type}`;
  }
}

function formatInput(
  transpiled: transpile.TranspiledParameter | transpile.TranspiledProperty
): string {
  return `${transpiled.name}: ${transpiled.typeReference.raw}${
    transpiled.optional ? " = None" : ""
  }`;
}

function formatInvocation(
  type: transpile.TranspiledType,
  inputs: string[],
  method?: string
) {
  let target = type.fqn;
  if (method) {
    target = `${target}.${method}`;
  }
  return `${target}(${formatInputs(inputs, 1 + target.length)})`;
}

function formatImport(type: transpile.TranspiledType) {
  return `import ${type.moduleFqn}`;
}

function formatSignature(name: string, inputs: string[]) {
  const def = "def ";
  // length of the word 'def' +
  // length of the method name +
  // 1 opening paranthesis
  const indent = def.length + name.length + 1;
  return `${def}${name}(${formatInputs(inputs, indent)})`;
}

function formatInputs(inputs: string[], indent: number) {
  return inputs.join(`, \n${" ".repeat(indent)}`);
}

/**
 * Hack to convert a jsii property to a parameter for
 * python specific parameter expansion.
 */
function propertyToParameter(
  callable: reflect.Callable,
  property: reflect.Property
): reflect.Parameter {
  return {
    docs: property.docs,
    method: callable,
    name: property.name,
    optional: property.optional,
    parentType: property.parentType,
    spec: property.spec,
    system: property.system,
    type: property.type,
    variadic: false,
  };
}
