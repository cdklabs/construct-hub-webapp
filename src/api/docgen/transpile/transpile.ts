import * as reflect from "jsii-reflect";

/**
 * Outcome of transpiling a jsii struct.
 */
export interface TranspiledStruct {
  /**
   * The (transpiled) type.
   */
  readonly type: TranspiledType;
  /**
   * The name.
   */
  readonly name: string;
  /**
   * The import statement needed in order to use this struct.
   */
  readonly import: string;
  /**
   * How to initialize this struct.
   */
  readonly initialization: string;
}

/**
 * Outcome of transpiling a jsii class.
 */
export interface TranspiledClass {
  /**
   * The (transpiled) type.
   */
  readonly type: TranspiledType;
  /**
   * The name.
   */
  readonly name: string;
}

/**
 * Outcome of transpiling a jsii callable.
 */
export interface TranspiledCallable {
  /**
   * The (transpiled) parent type.
   */
  readonly parentType: TranspiledType;
  /**
   * How a signature of the callable looks like.
   */
  readonly signature: string;
  /**
   * The name.
   */
  readonly name: string;
  /**
   * The import statement needed in order to use this callable.
   */
  readonly import: string;
  /**
   * How an invocation of this callable looks like.
   */
  readonly invocation: string;
  /**
   * The jsii parameters this callable accepts.
   */
  readonly parameters: reflect.Parameter[];
}

/**
 * Outcome of transpiling a jsii parameter.
 */
export interface TranspiledParameter {
  /**
   * The name.
   */
  readonly name: string;
  /**
   * The (transpiled) parent type.
   */
  readonly parentType: TranspiledType;
  /**
   * The (transpiled) type reference.
   */
  readonly typeReference: TranspiledTypeReference;
  /**
   * Whether or not the parameter is optional.
   */
  readonly optional: boolean;
}

/**
 * Outcome of transpiling a jsii interface (not data type).
 */
export interface TranspiledInterface {
  /**
   * The (transpiled) type.
   */
  readonly type: TranspiledType;
  /**
   * The name.
   */
  readonly name: string;
}

/**
 * Outcome of transpiling a generic jsii type.
 */
export interface TranspiledType {
  /**
   * The language specific fqn.
   */
  readonly fqn: string;
  /**
   * The language specific module fqn (mode + submodule).
   */
  readonly moduleFqn: string;
  /**
   * Simple name of the type.
   */
  readonly name: string;
}

/**
 * Options for how to render a string representation of a type reference.
 */
export interface TranspiledTypeReferenceToStringOptions {
  /**
   * Type formatter.
   */
  typeFormatter?: (t: TranspiledType) => string;
  /**
   * String formatter.
   */
  stringFormatter?: (formatter: string) => string;
}

/**
 * Outcome of transpiling a jsii type reference.
 */
export class TranspiledTypeReference {
  /**
   * Create a type reference that reprensents a primitive.
   */
  public static primitive(
    transpile: Transpile,
    ref: reflect.TypeReference,
    primitive: string
  ): TranspiledTypeReference {
    return new TranspiledTypeReference(transpile, ref, primitive);
  }
  /**
   * Create a type reference that represents any type.
   */
  public static any(
    transpile: Transpile,
    ref: reflect.TypeReference
  ): TranspiledTypeReference {
    return new TranspiledTypeReference(transpile, ref, undefined, true);
  }
  /**
   * Create a type reference that reprenets a concrete type.
   */
  public static type(
    transpile: Transpile,
    ref: reflect.TypeReference,
    type: TranspiledType
  ): TranspiledTypeReference {
    return new TranspiledTypeReference(
      transpile,
      ref,
      undefined,
      undefined,
      type
    );
  }
  /**
   * Create a type reference that reprenets an array of a type reference.
   */
  public static arrayOfType(
    transpile: Transpile,
    ref: reflect.TypeReference,
    tf: TranspiledTypeReference
  ): TranspiledTypeReference {
    return new TranspiledTypeReference(
      transpile,
      ref,
      undefined,
      undefined,
      undefined,
      tf
    );
  }
  /**
   * Create a type reference that reprenets a map of a type reference.
   */
  public static mapOfType(
    transpile: Transpile,
    ref: reflect.TypeReference,
    tf: TranspiledTypeReference
  ): TranspiledTypeReference {
    return new TranspiledTypeReference(
      transpile,
      ref,
      undefined,
      undefined,
      undefined,
      undefined,
      tf
    );
  }
  /**
   * Create a type reference that reprenets a union of a type references.
   */
  public static unionOfTypes(
    transpile: Transpile,
    ref: reflect.TypeReference,
    tfs: TranspiledTypeReference[]
  ): TranspiledTypeReference {
    return new TranspiledTypeReference(
      transpile,
      ref,
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
      tfs
    );
  }

  private constructor(
    /**
     * A transpiler
     */
    private readonly transpile: Transpile,
    /**
     * The original type reference.
     */
    private readonly ref: reflect.TypeReference,
    /**
     * Primitive type ref.
     */
    private readonly primitive?: string,
    /**
     * 'Any' type ref
     */
    private readonly isAny?: boolean,
    /**
     * Concrete type.
     */
    private readonly type?: TranspiledType,
    /**
     * Array of ref.
     */
    private readonly arrayOfType?: TranspiledTypeReference,
    /**
     * Map of ref.
     */
    private readonly mapOfType?: TranspiledTypeReference,
    /**
     * Union of ref.
     */
    private readonly unionOfTypes?: TranspiledTypeReference[]
  ) {}

  public toString(options?: TranspiledTypeReferenceToStringOptions): string {
    const tFormatter = options?.typeFormatter ?? ((t) => t.fqn);
    const sFormatter = options?.stringFormatter ?? ((s) => s);
    if (this.primitive) {
      return sFormatter(this.primitive);
    }
    if (this.type) {
      return tFormatter(this.type);
    }
    if (this.isAny) {
      return sFormatter(this.transpile.any());
    }
    if (this.arrayOfType) {
      const ref = this.arrayOfType.toString(options);
      return this.transpile.listOf(ref);
    }
    if (this.mapOfType) {
      const ref = this.mapOfType.toString(options);
      return this.transpile.mapOf(ref);
    }
    if (this.unionOfTypes) {
      const refs = this.unionOfTypes.map((t) => t.toString(options));
      return this.transpile.unionOf(refs);
    }
    throw new Error(`Invalid type reference: ${this.ref.toString()}`);
  }
}

/**
 * Outcome of transpiling a jsii property.
 */
export type TranspiledProperty = TranspiledParameter;

/**
 * Outcome of transpiling a jsii enum.
 */
export interface TranspiledEnum {
  /**
   * The language specific fqn.
   */
  readonly fqn: string;
  /**
   * The name.
   */
  readonly name: string;
}

/**
 * Outcome of transpiling a specific enum member.
 */
export interface TranspiledEnumMember {
  /**
   * The language specific fqn.
   */
  readonly fqn: string;
  /**
   * The name.
   */
  readonly name: string;
}

/**
 * Language transpiling for jsii types.
 */
export interface Transpile {
  /**
   * The language of the tranpiler.
   */
  language: string;

  /**
   * Transpile a callable (method, static function, initializer)
   */
  callable(callable: reflect.Callable): TranspiledCallable;

  /**
   * Transpile a class.
   */
  class(klass: reflect.ClassType): TranspiledClass;

  /**
   * Transpile a struct (data type interface).
   */
  struct(struct: reflect.InterfaceType): TranspiledStruct;

  /**
   * Transpile an interface (non data type).
   */
  interface(iface: reflect.InterfaceType): TranspiledInterface;

  /**
   * Transpile a parameter.
   */
  parameter(parameter: reflect.Parameter): TranspiledParameter;

  /**
   * Transpile a property.
   */
  property(property: reflect.Property): TranspiledProperty;

  /**
   * Transpile an enum.
   */
  enum(enu: reflect.EnumType): TranspiledEnum;

  /**
   * Transpile an enum member.
   */
  enumMember(em: reflect.EnumMember): TranspiledEnumMember;

  /**
   * Transpile a type.
   */
  type(type: reflect.Type): TranspiledType;

  /**
   * Transpile (recursively) a type reference.
   */
  typeReference(typeReference: reflect.TypeReference): TranspiledTypeReference;

  /**
   * How a union looks like in the target language.
   */
  unionOf(types: string[]): string;

  /**
   * How a list looks like in the target language.
   */
  listOf(type: string): string;

  /**
   * How a map looks like in the target language.
   */
  mapOf(type: string): string;

  /**
   * How the 'any' type is represented in the target language.
   */
  any(): string;

  /**
   * How the 'string' type is represented in the target language.
   */
  str(): string;

  /**
   * How the 'number' type is represented in the target language.
   */
  number(): string;

  /**
   * How the 'boolean' type is represented in the target language.
   */
  boolean(): string;

  /**
   * How the 'json' type is represented in the target language.
   */
  json(): string;

  /**
   * How the 'date' type is represented in the target language.
   */
  date(): string;

  /**
   * How a readme is displayed in the target language.
   */
  readme(readme: string): string;
}

// interface merging so we don't have to implement these methods
// in the abstract class.
export interface TranspileBase extends Transpile {}

/**
 * Common functionality between different transpilers.
 */
export abstract class TranspileBase implements Transpile {
  constructor(public readonly language: string) {}

  public typeReference(ref: reflect.TypeReference): TranspiledTypeReference {
    if (ref.type) {
      const transpiled = this.type(ref.type);
      return TranspiledTypeReference.type(this, ref, transpiled);
    }

    if (ref.unionOfTypes) {
      const transpiled = ref.unionOfTypes.map((t) => this.typeReference(t));
      return TranspiledTypeReference.unionOfTypes(this, ref, transpiled);
    }

    if (ref.arrayOfType) {
      const transpiled = this.typeReference(ref.arrayOfType);
      return TranspiledTypeReference.arrayOfType(this, ref, transpiled);
    }

    if (ref.mapOfType) {
      const transpiled = this.typeReference(ref.mapOfType);
      return TranspiledTypeReference.mapOfType(this, ref, transpiled);
    }

    if (ref.isAny) {
      return TranspiledTypeReference.any(this, ref);
    }

    if (ref.primitive) {
      let transpiled;
      switch (ref.primitive) {
        case "string":
          transpiled = this.str();
          break;
        case "boolean":
          transpiled = this.boolean();
          break;
        case "number":
          transpiled = this.number();
          break;
        case "date":
          transpiled = this.date();
          break;
        case "json":
          transpiled = this.json();
          break;
        default:
          throw new Error(`Unsupported primitive type '${ref.primitive}'`);
      }
      return TranspiledTypeReference.primitive(this, ref, transpiled);
    }

    throw new Error(`Unsupported type: ${ref.toString()}`);
  }
}
