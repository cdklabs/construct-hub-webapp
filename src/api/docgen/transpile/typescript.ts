import * as reflect from "jsii-reflect";
import * as transpile from "./transpile";

/**
 * A TypeScript transpiler.
 */
export class TypeScriptTranspile extends transpile.TranspileBase {
  constructor(private readonly ts: reflect.TypeSystem) {
    super("typescript");
  }

  public readme(readme: string): string {
    return readme;
  }

  public unionOf(types: string[]): string {
    return `${types.join(" | ")}`;
  }

  public listOf(type: string): string {
    return `${type}[]`;
  }

  public mapOf(type: string): string {
    return `{[ key: string ]: ${type}}`;
  }

  public any(): string {
    return "any";
  }

  public boolean(): string {
    return "boolean";
  }

  public str(): string {
    return "string";
  }

  public number(): string {
    return `number`;
  }

  public date(): string {
    return "Date";
  }

  public json(): string {
    return "object";
  }

  public enum(enu: reflect.EnumType): transpile.TranspiledEnum {
    return {
      fqn: this.type(enu).fqn,
      name: enu.name,
    };
  }

  public enumMember(em: reflect.EnumMember): transpile.TranspiledEnumMember {
    return {
      fqn: `${this.enum(em.enumType).fqn}.${em.name}`,
      name: em.name,
    };
  }

  public property(property: reflect.Property): transpile.TranspiledProperty {
    return {
      name: property.name,
      parentType: this.type(property.parentType),
      typeReference: this.typeReference(property.type),
      optional: property.optional,
    };
  }

  public class(klass: reflect.ClassType): transpile.TranspiledClass {
    return {
      name: klass.name,
      type: this.type(klass),
    };
  }

  public parameter(
    parameter: reflect.Parameter
  ): transpile.TranspiledParameter {
    return {
      name: parameter.name,
      parentType: this.type(parameter.parentType),
      typeReference: this.typeReference(parameter.type),
      optional: parameter.optional,
    };
  }

  public struct(struct: reflect.InterfaceType): transpile.TranspiledStruct {
    const type = this.type(struct);
    const inputs = struct.allProperties.map((p) =>
      this.formatInput(this.property(p))
    );
    return {
      type: type,
      name: struct.name,
      import: formatImport(type),
      initialization: formatInitialization(type, inputs),
    };
  }

  public callable(callable: reflect.Callable): transpile.TranspiledCallable {
    const type = this.type(callable.parentType);
    const parameters = callable.parameters.sort(this.optionalityCompare);
    const name = callable.name;
    const inputs = parameters.map((p) => this.formatInput(this.parameter(p)));

    const invocation =
      callable.kind === reflect.MemberKind.Initializer
        ? formatInitialization(type, inputs)
        : formatInvocation(type, inputs, name);

    return {
      name,
      parentType: type,
      import: formatImport(type),
      parameters,
      signature: formatSignature(name, inputs),
      invocation,
    };
  }

  public interface(
    iface: reflect.InterfaceType
  ): transpile.TranspiledInterface {
    return {
      name: iface.name,
      type: this.type(iface),
    };
  }

  // TODO this method is a mess, refactor at some point...
  public type(type: reflect.Type): transpile.TranspiledType {
    const t = this.ts.findFqn(type.fqn);
    const submodule = this.findSubmodule(t);
    const moduleFqn = submodule ? submodule.fqn : t.assembly.name;
    const submoduleName = submodule?.name;
    const moduleName = t.assembly.name;

    const fqn = [moduleName];

    if (type.namespace) {
      fqn.push(type.namespace);
    }

    fqn.push(type.name);

    return {
      fqn: fqn.join("."),
      moduleFqn,
      name: type.name,
      module: moduleName,
      submodule: submoduleName,
    };
  }

  private optionalityCompare(
    p1: reflect.Parameter,
    p2: reflect.Parameter
  ): number {
    if (!p1.optional && p1.optional) {
      return -1;
    }
    if (!p2.optional && p1.optional) {
      return 1;
    }
    return 0;
  }

  private formatInput(
    transpiled: transpile.TranspiledParameter | transpile.TranspiledProperty
  ): string {
    const tf = transpiled.typeReference.toString({
      typeFormatter: (t) => t.name,
    });
    return `${transpiled.name}${transpiled.optional ? "?" : ""}: ${tf}`;
  }
}

function formatInitialization(
  type: transpile.TranspiledType,
  inputs: string[]
) {
  const target = type.submodule ? `${type.submodule}.${type.name}` : type.name;
  return `new ${target}(${formatInputs(inputs)})`;
}

function formatInvocation(
  type: transpile.TranspiledType,
  inputs: string[],
  method: string
) {
  const target = type.submodule ? `${type.submodule}.${type.name}` : type.name;
  return `${target}.${method}(${formatInputs(inputs)})`;
}

function formatImport(type: transpile.TranspiledType) {
  if (type.submodule) {
    return `import { ${type.submodule} } from '${type.module}'`;
  } else {
    return `import { ${type.name} } from '${type.module}'`;
  }
}

function formatSignature(name: string, inputs: string[]) {
  return `public ${name}(${formatInputs(inputs)})`;
}

function formatInputs(inputs: string[]) {
  return inputs.join(", ");
}
