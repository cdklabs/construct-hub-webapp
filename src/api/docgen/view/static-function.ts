import * as reflect from "jsii-reflect";
import { Markdown } from "../render/markdown";
import { Transpile } from "../transpile/transpile";
import { Parameter } from "./parameter";
import { View } from "./view";

export class StaticFunction implements View {
  constructor(
    private readonly transpile: Transpile,
    private readonly method: reflect.Method
  ) {}

  public get markdown(): Markdown {
    const transpiled = this.transpile.callable(this.method);

    const md = new Markdown({
      header: {
        title: transpiled.name,
        code: true,
        deprecated: this.method.docs.deprecated,
      },
    });

    md.code(
      this.transpile.language,
      `${transpiled.requirement}`,
      "",
      `${transpiled.invocation}`
    );

    for (const parameter of transpiled.parameters) {
      md.section(new Parameter(this.transpile, parameter).markdown);
    }

    return md;
  }
}
