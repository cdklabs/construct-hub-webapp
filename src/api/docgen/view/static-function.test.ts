import * as reflect from "jsii-reflect";
import { PythonTranspile } from "../transpile/python";
import { TypeScriptTranspile } from "../transpile/typescript";
import { StaticFunction } from "./static-function";

const assembly: reflect.Assembly = (global as any).assembly;

describe("python", () => {
  const transpile = new PythonTranspile();
  test("snapshot", () => {
    const staticFunction = new StaticFunction(transpile, findStaticFunction());
    expect(staticFunction.render().render()).toMatchSnapshot();
  });
});

describe("typescript", () => {
  const transpile = new TypeScriptTranspile();
  test("snapshot", () => {
    const staticFunction = new StaticFunction(transpile, findStaticFunction());
    expect(staticFunction.render().render()).toMatchSnapshot();
  });
});

function findStaticFunction(): reflect.Method {
  for (const klass of assembly.system.classes) {
    for (const method of klass.ownMethods) {
      if (method.static) {
        return method;
      }
    }
  }
  throw new Error("Assembly does not contain a static function");
}
