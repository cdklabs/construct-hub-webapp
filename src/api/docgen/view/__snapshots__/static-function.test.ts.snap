// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`python snapshot 1`] = `
" \`grant_read\` <span data-heading-title=\\"\`grant_read\`\\" data-heading-id=\\"awscdkawsecrauthorizationtokengrantread\\"></span>

\`\`\`python
import aws_cdk.aws_ecr

aws_cdk.aws_ecr.AuthorizationToken.grant_read(grantee: IGrantable)
\`\`\`

# \`grantee\`<sup>Required</sup> <span data-heading-title=\\"\`grantee\`<sup>Required</sup>\\" data-heading-id=\\"awscdkawsecrauthorizationtokengrantee\\"></span>

- *Type:* [\`aws_cdk.aws_iam.IGrantable\`](#aws_cdk.aws_iam.IGrantable)

---
"
`;

exports[`typescript snapshot 1`] = `
" \`grantRead\` <span data-heading-title=\\"\`grantRead\`\\" data-heading-id=\\"awscdkawsecrauthorizationtokengrantread\\"></span>

\`\`\`typescript
import { AuthorizationToken } from '@aws-cdk/aws-ecr'

AuthorizationToken.grantRead(grantee: IGrantable)
\`\`\`

# \`grantee\`<sup>Required</sup> <span data-heading-title=\\"\`grantee\`<sup>Required</sup>\\" data-heading-id=\\"awscdkawsecrauthorizationtokengrantee\\"></span>

- *Type:* [\`@aws-cdk/aws-iam.IGrantable\`](#@aws-cdk/aws-iam.IGrantable)

---
"
`;
